akka-sample-persistence-typed-scala/src/main/scala/sample/persistence/ShoppingCart.scala

```diff
@@ -0,0 +1,119 @@
+    case class State(items: Map[String, Int], checkedOut: Boolean) {
+        def updateItem(productId: String, quantity: Int) = {
+            quantity match {
+            case 0 => copy(items = items - productId)
+            case _ => copy(items = items + (productId -> quantity))
+            }
+        }
+
+        def checkout = copy(checkedOut = true)
+    }
+    object State {
+        val empty = State(Map.empty, false)
+    }
+    
+
+    /**
+     * This interface defines all the commands that the ShoppingCart persistent actor supports.
+    */
+    sealed trait Command
+
+    sealed trait Result
+    case object OK extends Result
+    case class Failed(msg: String) extends Result
+
+    /**
+     * A command to update an item.
+    *
+    * It can reply with `Done`, which is sent back to the caller when
+    * all the events emitted by this command are successfully persisted.
+    */
+    case class UpdateItem(productId: String, quantity: Int, replyTo: ActorRef[Result]) extends Command with ExpectingReply[Result]
+
+    /**
+     * A command to get the current state of the shopping cart.
+    *
+    * It can reply with the state of the ShoppingCart
+    */
+    case class Get(replyTo: ActorRef[State]) extends Command with ExpectingReply[State]
+
+    /**
+     * A command to checkout the shopping cart.
+    *
+    * It can reply with Done, which will be returned when the fact that checkout
+    * has started has successfully been persisted.
+    */
+    case class Checkout(replyTo: ActorRef[Result]) extends Command with ExpectingReply[Result]
+
+    /**
+     * This interface defines all the events that the ShoppingCart supports.
+     */
+    sealed trait Event
+
+    /**
+     * An event that represents a item updated event.
+     */
+    case class ItemUpdated(productId: String, quantity: Int) extends Event
+
+    /**
+     * An event that represents a checked out event.
+     */
+    case object CheckedOut extends Event
+
+    def behavior(id: String): Behavior[Command] =
+        EventSourcedBehavior[Command, Event, State](
+            PersistenceId(id),
+            State.empty,
+            (state, command) => 
+                if (state.checkedOut) checkedOutShoppingCart(state, command)
+                else openShoppingCart(state, command),
+            (state, event) => event match {
+                case ItemUpdated(productId, quantity) => state.updateItem(productId, quantity)
+                case CheckedOut => state.checkout
+            }
+        )
+
+    def openShoppingCart(state: State, command: Command): Effect[Event, State] = command match {
+        case cmd @ UpdateItem(_, quantity, _) if quantity < 0 => 
+            Effect.reply(cmd)(Failed("Quantity must be greater than zero"))
+        case cmd @ UpdateItem(productId, 0, _) if !state.items.contains(productId) =>
+            Effect.reply(cmd)(Failed("Cannot delete item that is not already in cart"))
+        case cmd @ UpdateItem(productId, quantity, _) =>
```

Comment:
I prefer one case of all `UpdateItem`, and then if statements inside. Less boilerplate but also sealed exhaustiveness check remains